File: apps/worker/src/worker/jobs/export_docx.py
from __future__ import annotations

from typing import Any

from worker.jobs.common import (
    RetryableJobError,
    UnrecoverableJobError,
    failure_strategy,
)
from worker.jobs.render_docx import render_document_value_to_docx_bytes

DOCX_CONTENT_TYPE = "application/vnd.openxmlformats-officedocument.wordprocessingml.document"


class ExportDocxProcessor:
    def __init__(self, repository: Any, storage: Any, events: Any) -> None:
        self._repository = repository
        self._storage = storage
        self._events = events

    def process(
        self,
        payload: dict[str, object],
        *,
        job_id: str,
        attempts_made: int,
        max_attempts: int,
    ) -> None:
        row = self._repository.get_export_job_row(str(payload["exportId"]))
        final_attempt = attempts_made + 1 >= max(1, max_attempts)

        if not row:
            self._emit_failure(
                payload,
                job_id=job_id,
                code="export_not_found",
                message=f"Export {payload['exportId']} not found",
                attempts_made=attempts_made,
                max_attempts=max_attempts,
                terminal=True,
                will_retry=False,
            )
            raise UnrecoverableJobError(f"Export {payload['exportId']} not found")

        if (
            row["user_id"] != payload["userId"]
            or row["document_id"] != payload["documentId"]
            or row["workspace_id"] != payload["workspaceId"]
        ):
            self._fail_job(
                payload,
                job_id=job_id,
                error_code="payload_mismatch",
                error_message="Job payload does not match export record",
                retryable=False,
                attempts_made=attempts_made,
                max_attempts=max_attempts,
            )

        if row["status"] == "ready":
            return

        started = self._repository.mark_processing(str(payload["exportId"]), job_id)
        if not started:
            again = self._repository.get_export_job_row(str(payload["exportId"]))
            if again and again["status"] == "ready":
                return
            if again and again["status"] == "processing" and again["bullmq_job_id"] == job_id:
                pass
            elif again and again["status"] == "processing":
                return
            message = f"Could not move export {payload['exportId']} to processing"
            self._emit_failure(
                payload,
                job_id=job_id,
                code="processing_transition_failed",
                message=message,
                attempts_made=attempts_made,
                max_attempts=max_attempts,
                terminal=final_attempt,
                will_retry=not final_attempt,
            )
            if final_attempt:
                self._repository.mark_failed(
                    str(payload["exportId"]),
                    error_code="processing_transition_failed",
                    error_message=message,
                )
                raise UnrecoverableJobError(message)
            raise RetryableJobError(message)

        document = self._repository.get_document(str(payload["documentId"]))
        if not document:
            self._fail_job(
                payload,
                job_id=job_id,
                error_code="document_missing",
                error_message="Document not found",
                retryable=False,
                attempts_made=attempts_made,
                max_attempts=max_attempts,
            )

        try:
            buffer = render_document_value_to_docx_bytes(
                title=str(document["title"]),
                value=list(document["content_json"]),
            )
            storage_key = self._storage.create_export_key(
                str(payload["userId"]),
                str(payload["exportId"]),
            )
            self._storage.write_export_file(storage_key, buffer)
            ready = self._repository.mark_ready(
                str(payload["exportId"]),
                storage_key=storage_key,
                content_type=DOCX_CONTENT_TYPE,
                file_size_bytes=len(buffer),
            )
            if not ready.get("ok"):
                self._fail_job(
                    payload,
                    job_id=job_id,
                    error_code="ready_transition_failed",
                    error_message="Could not finalize export as ready",
                    retryable=False,
                    attempts_made=attempts_made,
                    max_attempts=max_attempts,
                )
            self._events.log_launch_event(
                "export.docx_ready",
                {
                    "exportId": payload["exportId"],
                    "userId": payload["userId"],
                    "documentId": payload["documentId"],
                    "fileSizeBytes": len(buffer),
                },
            )
        except (RetryableJobError, UnrecoverableJobError):
            raise
        except Exception as error:
            self._fail_job(
                payload,
                job_id=job_id,
                error_code="export_render_failed",
                error_message=str(error)[:2000],
                retryable=True,
                attempts_made=attempts_made,
                max_attempts=max_attempts,
            )

    def _fail_job(
        self,
        payload: dict[str, object],
        *,
        job_id: str,
        error_code: str,
        error_message: str,
        retryable: bool,
        attempts_made: int,
        max_attempts: int,
    ) -> None:
        strategy = failure_strategy(
            retryable=retryable,
            attempts_made=attempts_made,
            max_attempts=max_attempts,
        )
        self._emit_failure(
            payload,
            job_id=job_id,
            code=error_code,
            message=error_message,
            attempts_made=attempts_made,
            max_attempts=max_attempts,
            terminal=strategy["mark_failed"] or strategy["unrecoverable"],
            will_retry=retryable and not (strategy["mark_failed"] or strategy["unrecoverable"]),
        )
        if strategy["mark_failed"]:
            self._repository.mark_failed(
                str(payload["exportId"]),
                error_code=error_code,
                error_message=error_message,
            )
            self._events.log_launch_event(
                "export.docx_failed",
                {
                    "exportId": payload["exportId"],
                    "userId": payload["userId"],
                    "documentId": payload["documentId"],
                    "errorCode": error_code,
                    "attemptsMade": attempts_made + 1,
                    "maxAttempts": max_attempts,
                },
            )
        if strategy["unrecoverable"]:
            raise UnrecoverableJobError(error_message)
        raise RetryableJobError(error_message)

    def _emit_failure(
        self,
        payload: dict[str, object],
        *,
        job_id: str,
        code: str,
        message: str,
        attempts_made: int,
        max_attempts: int,
        terminal: bool,
        will_retry: bool,
    ) -> None:
        self._events.log_error_event(
            {
                "domain": "worker",
                "failureClass": "system",
                "code": code,
                "jobId": job_id,
                "exportId": payload["exportId"],
                "documentId": payload["documentId"],
                "userId": payload["userId"],
                "workspaceId": payload["workspaceId"],
                "message": message,
                "attemptNumber": attempts_made + 1,
                "maxAttempts": max_attempts,
                "terminal": terminal,
                "willRetry": will_retry,
                "workerJob": "export_docx",
            }
        )


__all__ = ["ExportDocxProcessor", "RetryableJobError", "UnrecoverableJobError"]

File: apps/worker/src/jobs/export-docx.ts
import { logLaunchEvent } from "@aqshara/observability";
import { createDatabase, documents } from "@aqshara/database";
import {
  getExportJobRow,
  markExportFailed,
  markExportProcessing,
  markExportReady,
} from "@aqshara/database/export-job";
import type { DocumentValue } from "@aqshara/documents";
import type { ExportDocxPayload } from "@aqshara/queue";
import { createStorageKey, writeExportFile } from "@aqshara/storage";
import { UnrecoverableError } from "bullmq";
import { eq } from "drizzle-orm";
import { renderDocumentValueToDocxBuffer } from "../render-docx.ts";
import { logWorkerExportFailureEvent } from "./export-docx-error-event.ts";
import { getExportFailureStrategy } from "./export-docx-failure-strategy.ts";

const DOCX_CONTENT_TYPE =
  "application/vnd.openxmlformats-officedocument.wordprocessingml.document";

async function failExportJob(input: {
  db: ReturnType<typeof createDatabase>;
  payload: ExportDocxPayload;
  bullmqJobId: string;
  errorCode: string;
  errorMessage: string;
  retryable: boolean;
  attemptsMade: number;
  maxAttempts: number;
}): Promise<never> {
  const strategy = getExportFailureStrategy({
    retryable: input.retryable,
    attemptsMade: input.attemptsMade,
    maxAttempts: input.maxAttempts,
  });
  const terminal = strategy.markFailed || strategy.unrecoverable;

  logWorkerExportFailureEvent({
    payload: input.payload,
    jobId: input.bullmqJobId,
    code: input.errorCode,
    message: input.errorMessage,
    attemptsMade: input.attemptsMade,
    maxAttempts: input.maxAttempts,
    terminal,
    willRetry: input.retryable && !terminal,
  });

  if (strategy.markFailed) {
    await markExportFailed(input.db, {
      exportId: input.payload.exportId,
      errorCode: input.errorCode,
      errorMessage: input.errorMessage,
    });

    logLaunchEvent("export.docx_failed", {
      exportId: input.payload.exportId,
      userId: input.payload.userId,
      documentId: input.payload.documentId,
      errorCode: input.errorCode,
      attemptsMade: input.attemptsMade + 1,
      maxAttempts: input.maxAttempts,
    });
  }

  if (strategy.unrecoverable) {
    throw new UnrecoverableError(input.errorMessage);
  }

  throw new Error(input.errorMessage);
}

async function markExportFailedBestEffort(input: {
  db: ReturnType<typeof createDatabase>;
  exportId: string;
  errorCode: string;
  errorMessage: string;
}): Promise<void> {
  try {
    await markExportFailed(input.db, {
      exportId: input.exportId,
      errorCode: input.errorCode,
      errorMessage: input.errorMessage,
    });
  } catch {
    // Preserve the original worker failure if DB reconciliation also fails.
  }
}

export async function processExportDocxJob(
  payload: ExportDocxPayload,
  bullmqJobId: string,
  attempt: { attemptsMade: number; maxAttempts: number } = {
    attemptsMade: 0,
    maxAttempts: 1,
  },
): Promise<void> {
  const db = createDatabase();
  const finalAttempt =
    attempt.attemptsMade + 1 >= Math.max(1, attempt.maxAttempts);

  let row;
  try {
    row = await getExportJobRow(db, payload.exportId);
  } catch (error) {
    const message =
      error instanceof Error ? error.message : "Failed to load export job";
    logWorkerExportFailureEvent({
      payload,
      jobId: bullmqJobId,
      code: "export_lookup_failed",
      message,
      attemptsMade: attempt.attemptsMade,
      maxAttempts: attempt.maxAttempts,
      terminal: finalAttempt,
      willRetry: !finalAttempt,
    });
    if (finalAttempt) {
      throw new UnrecoverableError(message);
    }
    throw error;
  }

  if (!row) {
    const message = `Export ${payload.exportId} not found`;
    logWorkerExportFailureEvent({
      payload,
      jobId: bullmqJobId,
      code: "export_not_found",
      message,
      attemptsMade: attempt.attemptsMade,
      maxAttempts: attempt.maxAttempts,
      terminal: true,
      willRetry: false,
    });
    throw new UnrecoverableError(message);
  }

  if (
    row.userId !== payload.userId ||
    row.documentId !== payload.documentId ||
    row.workspaceId !== payload.workspaceId
  ) {
    await failExportJob({
      db,
      payload,
      bullmqJobId,
      errorCode: "payload_mismatch",
      errorMessage: "Job payload does not match export record",
      retryable: false,
      attemptsMade: attempt.attemptsMade,
      maxAttempts: attempt.maxAttempts,
    });
  }

  if (row.status === "ready") {
    return;
  }

  let started;
  try {
    started = await markExportProcessing(db, {
      exportId: payload.exportId,
      bullmqJobId,
    });
  } catch (error) {
    const message =
      error instanceof Error ? error.message : "Failed to transition export";
    logWorkerExportFailureEvent({
      payload,
      jobId: bullmqJobId,
      code: "processing_transition_failed",
      message,
      attemptsMade: attempt.attemptsMade,
      maxAttempts: attempt.maxAttempts,
      terminal: finalAttempt,
      willRetry: !finalAttempt,
    });
    if (finalAttempt) {
      await markExportFailedBestEffort({
        db,
        exportId: payload.exportId,
        errorCode: "processing_transition_failed",
        errorMessage: message,
      });
      throw new UnrecoverableError(message);
    }
    throw error;
  }

  if (!started) {
    let again;
    try {
      again = await getExportJobRow(db, payload.exportId);
    } catch (error) {
      const message =
        error instanceof Error ? error.message : "Failed to reload export job";
      logWorkerExportFailureEvent({
        payload,
        jobId: bullmqJobId,
        code: "processing_transition_lookup_failed",
        message,
        attemptsMade: attempt.attemptsMade,
        maxAttempts: attempt.maxAttempts,
        terminal: finalAttempt,
        willRetry: !finalAttempt,
      });
      if (finalAttempt) {
        await markExportFailedBestEffort({
          db,
          exportId: payload.exportId,
          errorCode: "processing_transition_lookup_failed",
          errorMessage: message,
        });
        throw new UnrecoverableError(message);
      }
      throw error;
    }
    if (again?.status === "ready") {
      return;
    }
    if (again?.status === "processing" && again.bullmqJobId === bullmqJobId) {
      // BullMQ keeps the same job id across retry attempts.
    } else if (again?.status === "processing") {
      return;
    }
    const message = `Could not move export ${payload.exportId} to processing`;
    logWorkerExportFailureEvent({
      payload,
      jobId: bullmqJobId,
      code: "processing_transition_failed",
      message,
      attemptsMade: attempt.attemptsMade,
      maxAttempts: attempt.maxAttempts,
      terminal: finalAttempt,
      willRetry: !finalAttempt,
    });
    if (finalAttempt) {
      await markExportFailedBestEffort({
        db,
        exportId: payload.exportId,
        errorCode: "processing_transition_failed",
        errorMessage: message,
      });
      throw new UnrecoverableError(message);
    }
    throw new Error(message);
  }

  try {
    const [docRow] = await db
      .select()
      .from(documents)
      .where(eq(documents.id, payload.documentId))
      .limit(1);

    if (!docRow) {
      return await failExportJob({
        db,
        payload,
        bullmqJobId,
        errorCode: "document_missing",
        errorMessage: "Document not found",
        retryable: false,
        attemptsMade: attempt.attemptsMade,
        maxAttempts: attempt.maxAttempts,
      });
    }

    const value = docRow.contentJson as DocumentValue;
    const buffer = await renderDocumentValueToDocxBuffer({
      title: docRow.title,
      value,
    });

    const storageKey = createStorageKey(
      "exports",
      payload.userId,
      `${payload.exportId}.docx`,
    );

    await writeExportFile(storageKey, buffer);

    const ready = await markExportReady(db, {
      exportId: payload.exportId,
      storageKey,
      contentType: DOCX_CONTENT_TYPE,
      fileSizeBytes: buffer.length,
    });

    if (!ready.ok) {
      await failExportJob({
        db,
        payload,
        bullmqJobId,
        errorCode: "ready_transition_failed",
        errorMessage: "Could not finalize export as ready",
        retryable: false,
        attemptsMade: attempt.attemptsMade,
        maxAttempts: attempt.maxAttempts,
      });
    } else {
      logLaunchEvent("export.docx_ready", {
        exportId: payload.exportId,
        userId: payload.userId,
        documentId: payload.documentId,
        fileSizeBytes: buffer.length,
      });
    }
  } catch (error) {
    if (error instanceof UnrecoverableError) {
      throw error;
    }

    const message = error instanceof Error ? error.message : String(error);
    await failExportJob({
      db,
      payload,
      bullmqJobId,
      errorCode: "export_render_failed",
      errorMessage: message.slice(0, 2000),
      retryable: true,
      attemptsMade: attempt.attemptsMade,
      maxAttempts: attempt.maxAttempts,
    });
  }
}
